spv/backend.go

```diff
@@ -0,0 +1,301 @@
+// license that can be found in the LICENSE file.
+
+package spv
+
+import (
+	"context"
+	"runtime"
+	"sync"
+
+	"github.com/decred/dcrd/chaincfg/chainhash"
+	"github.com/decred/dcrd/dcrutil/v2"
+	"github.com/decred/dcrd/gcs"
+	"github.com/decred/dcrd/gcs/blockcf"
+	"github.com/decred/dcrd/txscript/v2"
+	"github.com/decred/dcrd/wire"
+	"github.com/decred/dcrwallet/errors/v2"
+	"github.com/decred/dcrwallet/p2p/v2"
+	"github.com/decred/dcrwallet/validate"
+	"github.com/decred/dcrwallet/wallet/v3"
+)
+
+var _ wallet.NetworkBackend = (*Syncer)(nil)
+
+// TODO: When using the Syncer as a NetworkBackend, keep track of in-flight
+// blocks and cfilters.  If one is already incoming, wait on that response.  If
+// that peer is lost, try a different peer.  Optionally keep a cache of fetched
+// data so it can be immediately returned without another call.
+
+func pickAny(*p2p.RemotePeer) bool { return true }
+
+// Blocks implements the Blocks method of the wallet.Peer interface.
+func (s *Syncer) Blocks(ctx context.Context, blockHashes []*chainhash.Hash) ([]*wire.MsgBlock, error) {
+	for {
+		if err := ctx.Err(); err != nil {
+			return nil, err
+		}
+		rp, err := s.pickRemote(pickAny)
+		if err != nil {
+			return nil, err
+		}
+		blocks, err := rp.Blocks(ctx, blockHashes)
+		if err != nil {
+			continue
+		}
+		return blocks, nil
+	}
+}
+
+// CFilters implements the CFilters method of the wallet.Peer interface.
+func (s *Syncer) CFilters(ctx context.Context, blockHashes []*chainhash.Hash) ([]*gcs.Filter, error) {
+	for {
+		if err := ctx.Err(); err != nil {
+			return nil, err
+		}
+		rp, err := s.pickRemote(pickAny)
+		if err != nil {
+			return nil, err
+		}
+		fs, err := rp.CFilters(ctx, blockHashes)
+		if err != nil {
+			continue
+		}
+		return fs, nil
+	}
+}
+
+// Headers implements the Headers method of the wallet.Peer interface.
+func (s *Syncer) Headers(ctx context.Context, blockLocators []*chainhash.Hash, hashStop *chainhash.Hash) ([]*wire.BlockHeader, error) {
+	for {
+		if err := ctx.Err(); err != nil {
+			return nil, err
+		}
+		rp, err := s.pickRemote(pickAny)
+		if err != nil {
+			return nil, err
+		}
+		hs, err := rp.Headers(ctx, blockLocators, hashStop)
+		if err != nil {
+			continue
+		}
+		return hs, nil
+	}
+}
+
+func (s *Syncer) String() string {
+	// This method is part of the wallet.Peer interface and will typically
+	// specify the remote address of the peer.  Since the syncer can encompass
+	// multiple peers, just use the qualified type as the string.
+	return "spv.Syncer"
+}
+
+// LoadTxFilter implements the LoadTxFilter method of the wallet.NetworkBackend
+// interface.
+func (s *Syncer) LoadTxFilter(ctx context.Context, reload bool, addrs []dcrutil.Address, outpoints []wire.OutPoint) error {
+	s.filterMu.Lock()
+	if reload || s.rescanFilter[s.filterWalletID] == nil {
+		s.rescanFilter[s.filterWalletID] = wallet.NewRescanFilter(nil, nil)
+		s.filterData[s.filterWalletID] = &blockcf.Entries{}
```

Comment:
The default value was previously `nil` rather than `blockcf.Entries{}`. Any reason for the change here?

Reply:
I changed `filterData` to a pointer because of this error `cannot call pointer method on s.filterData[s.filterWalletID]` due to it being in a map.